astroid/rebuilder.py

```diff
@@ -1383,14 +1393,32 @@ def visit_name(
 
     def visit_nonlocal(self, node: ast.Nonlocal, parent: NodeNG) -> nodes.Nonlocal:
         """Visit a Nonlocal node and return a new instance of it."""
-        return nodes.Nonlocal(
+        newnode = nodes.Nonlocal(
             names=node.names,
             lineno=node.lineno,
             col_offset=node.col_offset,
             end_lineno=node.end_lineno,
             end_col_offset=node.end_col_offset,
             parent=parent,
         )
+        names = set(newnode.names)
+        # Go through the tree and find where those names are created
+        scope = newnode
+        while len(names) != 0:
+            scope = scope.parent
+            if not scope:
+                # It's not inside a nested function or there are no variables with that name.
+                # Just ignore it as visit_global does when global is used in module scope.
+                break
```

Comment:
I just `break` if it's not inside a nested function, or there are no variables with that name. I'm not sure whether it's an acceptable behavior.

But I did this because of this similar case:

```python
x = 1
def f():
    x = 2
    global x
```

It should be `SyntaxError` in Python, but currently `astroid` could parse it without raising an exception.

So I think it's ok to do anything if there is a syntax error, so I just `break` here.

Reply:
Yes! We shouldn't crash on invalid code so this is the right approach